### PR TITLE
Android POC support

### PR DESCRIPTION
**Pull request by Fryzu**
Adds _proof-of-concept_ Android support:

* For now works on forked version of the [turbo android](https://github.com/hotwired/turbo-android/) library.
* Doesn't support multiple sessions yet.
* Doesn't support JS<->Native communication yet.

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -5,9 +5,10 @@ import org.apache.tools.ant.taskdefs.condition.Os
 buildscript {
     ext {
         buildToolsVersion = "31.0.0"
-        minSdkVersion = 21
+        minSdkVersion = 24
```
Comment (Fryzu) [inline android/build.gradle:8]: Based on https://github.com/hotwired/turbo-android/#requirements

File: .gitignore [hunk 2]
```diff
@@ -59,3 +59,5 @@ buck-out/
 # Ruby / CocoaPods
 /ios/Pods/
 /vendor/bundle/
+
+android/turbo
```
Comment (lukmccall) [inline .gitignore:63]:
```suggestion
android/turbo

```

File: android/app/src/main/java/com/hotwirerndemo/RNSession.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import android.util.Log
+import android.widget.FrameLayout
+import androidx.appcompat.app.AppCompatActivity
+import com.facebook.react.bridge.ReactContext
+import dev.hotwire.turbo.session.TurboSession
+import dev.hotwire.turbo.views.TurboWebView
+
+class RNSession(context: Context) : FrameLayout(context) {
+
+}
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNSession.kt:12]:
```suggestion
class RNSession(context: Context) : FrameLayout(context)

```

File: android/app/src/main/java/com/hotwirerndemo/RNSessionManager.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    override fun getName() = REACT_CLASS
+
+    companion object {
+        const val REACT_CLASS = "RNSession"
+    }
+
+    override fun createViewInstance(reactContext: ThemedReactContext) = RNSession(reactContext)
+
+}
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNSessionManager.kt:19]:
```suggestion

private const val REACT_CLASS = "RNSession"

class RNSessionManager(
    private val callerContext: ReactApplicationContext
) : SimpleViewManager<RNSession>() {
    override fun getName() = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext) = RNSession(reactContext)
}

```

File: android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            location = url,
+            destinationIdentifier = 1,
+            restoreWithCachedSnapshot = false,
+            reload = false,
+            callback = this,
+            identifier = url,
+            options = TurboVisitOptions()
+        )
+        session.visit(visit)
+        turboView.attachWebView(session.webView, { a: Boolean -> 0 })
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt:53]:
```suggestion
        turboView.attachWebView(session.webView) { 0 }
```

File: android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        val width = r - l
+        val height = b - t
+        session.webView.layout(0, 0, width, height)
+    }
+
+    private fun sendEvent(event: RNVisitableViewEvent, params: WritableMap) {
+        reactContext.getJSModule(RCTEventEmitter::class.java).receiveEvent(id, event.name, params)
+    }
+
+    // TurboSessionCallback =====
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt:70]:
```suggestion
    // region TurboSessionCallback
```

File: android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        return null
+    }
+
+    override fun formSubmissionStarted(location: String) {
+        Log.d("RNVisitableView", "formSubmissionStarted")
+    }
+
+    override fun formSubmissionFinished(location: String) {
+        Log.d("RNVisitableView", "formSubmissionFinished")
+    }
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNVisitableView.kt:147]:
```suggestion
    }
    
    // end region
```

File: android/app/src/main/java/com/hotwirerndemo/RNVisitableViewManager.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        )}.toMap()
+    }
+
+    companion object {
+        const val REACT_CLASS = "RNVisitableView"
+    }
+
+    override fun createViewInstance(reactContext: ThemedReactContext) = RNVisitableView(reactContext)
+
+}
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/RNVisitableViewManager.kt:41]:
```suggestion

private const val REACT_CLASS = "RNVisitableView"
      
class RNVisitableViewManager(
    private val callerContext: ReactApplicationContext
) : SimpleViewManager<RNVisitableView>() {

    override fun getName() = REACT_CLASS

    @ReactProp(name = "url")
    fun setUrl(view: RNVisitableView, url: String) {
        Log.d("RMVisitableView", "url set ${url}")
        view.openUrl(url)
    }

    override fun getExportedCustomBubblingEventTypeConstants(): Map<String, Any> {
        return RNVisitableViewEvent.values().map { it.name to mapOf(
            "phasedRegistrationNames" to mapOf(
                "bubbled" to it.jsCallbackName
            )
        )}.toMap()
    }

    override fun createViewInstance(reactContext: ThemedReactContext) = RNVisitableView(reactContext)
}

```

File: android/app/src/main/java/com/hotwirerndemo/ReactAppPackage.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+import com.facebook.react.bridge.NativeModule
+import com.facebook.react.bridge.ReactApplicationContext
+
+class ReactAppPackage: ReactPackage {
+
+    override fun createViewManagers(reactContext: ReactApplicationContext) = listOf(RNSessionManager(reactContext), RNVisitableViewManager(reactContext)).toMutableList()
+
+    override fun createNativeModules(reactContext: ReactApplicationContext): MutableList<NativeModule> = mutableListOf()
+
+}
```
Comment (lukmccall) [inline android/app/src/main/java/com/hotwirerndemo/ReactAppPackage.kt:13]:
```suggestion
}

```

File: android/settings.gradle [hunk 10]
```diff
@@ -6,4 +6,4 @@ includeBuild('../node_modules/react-native-gradle-plugin')
 if (settings.hasProperty("newArchEnabled") && settings.newArchEnabled == "true") {
     include(":ReactAndroid")
     project(":ReactAndroid").projectDir = file('../node_modules/react-native/ReactAndroid')
-}
+}
```
Comment (lukmccall) [inline android/settings.gradle:9]:
```suggestion
}

```

**Review by lukmccall — APPROVED** [review]

File: android/app/src/main/java/com/hotwirerndemo/MainApplication.java [hunk 11]
```diff
@@ -24,11 +25,12 @@ public boolean getUseDeveloperSupport() {
...
-          // Packages that cannot be autolinked yet can be added manually here, for example:
-          // packages.add(new MyReactNativePackage());
-          return packages;
+            @SuppressWarnings("UnnecessaryLocalVariable")
+            List<ReactPackage> packages = new PackageList(this).getPackages();
+            // Packages that cannot be autolinked yet can be added manually here, for example:
+            // packages.add(new MyReactNativePackage());
+            packages.add(new ReactAppPackage());
+            return packages;
         }
```
Comment (pfeiffer) [inline android/app/src/main/java/com/hotwirerndemo/MainApplication.java:34]: Could be reformatted so the actual change is more visible in the diff (ie the `packages.add(new ReactAppPackage());` line only)

**Pull request merged**